File: c/demo/hand_demo.cpp
#include "../common/revo3_common.h"

#include <cstdio>

namespace {

void print_first_values(const char *label, const float *values, int count) {
  std::printf("%s:", label);
  for (int i = 0; i < count; ++i) {
    std::printf(" %.2f", values[i]);
  }
  std::printf("\n");
}

void set_all(float *values, float value) {
  for (int i = 0; i < 21; ++i) {
    values[i] = value;
  }
}


} // namespace

int main(int argc, char **argv) {
  init_logging(LOG_LEVEL_INFO);

  Revo3Context ctx;
  if (!revo3_init_from_args(ctx, argc, argv)) {
    return 1;
  }

  std::printf("=== Revo3 C++ Hand Demo ===\n");
  revo3_print_device_info(ctx.handle, ctx.slave_id);

  CRevo3MotorStatusData *status = revo3_get_motor_status_data(ctx.handle, ctx.slave_id);
  if (status) {
    print_first_values("Initial positions[0..4]", status->positions, 5);
    print_first_values("Initial velocities[0..4]", status->velocities, 5);
    print_first_values("Initial currents[0..4]", status->currents, 5);
    free_revo3_motor_status_data(status);
  }

  uint16_t enabled = revo3_get_all_touch_modules_enabled(ctx.handle, ctx.slave_id);
  if (enabled != 0) {
    uint16_t summary[42] = {0};
    if (revo3_get_touch_summary(ctx.handle, ctx.slave_id, summary) == 0) {
      std::printf("Touch summary[0..7]:");
      for (int i = 0; i < 8; ++i) {
        std::printf(" %u", summary[i]);
      }
      std::printf("\n");
    }
  } else {
    std::printf("Touch modules are not enabled or not available on this device.\n");
  }

  float zeros[21] = {0.0f};
  set_all(zeros, 0.0f);

  // 1. Explicitly switch all 21 joints to MIT/Impedance mode (4)
  std::printf("Switching all joints to Impedance control mode (4) for smooth trajectory planning...\n");
  revo3_multi_joint_control(ctx.handle, ctx.slave_id, 4, zeros);
  revo3_sleep_ms(100);

  float grip_targets[21] = {0.0f};
  set_all(grip_targets, 0.0f);
  for (int joint : {1, 2, 5, 6, 9, 10, 13, 14}) {
    grip_targets[joint] = 60.0f;
  }
  for (int joint : {17, 18, 19, 20}) {
    grip_targets[joint] = 45.0f;
  }

  // =========================================================================
  // Step 1: Initial Stretch (First open all joints to 0 deg with smooth trajectory)
  // =========================================================================
  std::printf("Step 1: Open all fingers to 0 deg (Initial Stretch via revo3_move_hand_wait)...\n");
  revo3_move_hand_wait(ctx.handle, ctx.slave_id, zeros, 21, 0.7f, 0.025f);

  // =========================================================================
  // Step 2: Grip (Move whole hand to 60 deg with smooth trajectory)
  // =========================================================================
  std::printf("Step 2: Close whole hand to 60 deg (Grip via revo3_move_hand_wait)...\n");
  revo3_move_hand_wait(ctx.handle, ctx.slave_id, grip_targets, 21, 0.7f, 0.025f);

  // =========================================================================
  // Step 3: Open again (Return whole hand to 0 deg with smooth trajectory)
  // =========================================================================
  std::printf("Step 3: Open whole hand to 0 deg (Release via revo3_move_hand_wait)...\n");
  revo3_move_hand_wait(ctx.handle, ctx.slave_id, zeros, 21, 0.7f, 0.025f);

  // =========================================================================
  // Step 2: Test fingers one by one (Pinky -> Ring -> Middle -> Index -> Thumb)
  // =========================================================================
  struct FingerTest {
    const char *name;
    uint16_t idx;
    float target[4];
  };

  FingerTest diagnostic_sequence[] = {
    {"Pinky (Little Finger)", 4, {0.0f, 45.0f, 45.0f, 0.0f}},
    {"Ring Finger", 3, {0.0f, 45.0f, 45.0f, 0.0f}},
    {"Middle Finger", 2, {0.0f, 45.0f, 45.0f, 0.0f}},
    {"Index Finger", 1, {0.0f, 45.0f, 45.0f, 0.0f}}
  };

  float finger_zero[4] = {0.0f};

  for (int f = 0; f < 4; ++f) {
    std::printf("Testing: %s (smooth trajectory)...\n", diagnostic_sequence[f].name);
    // Bend finger
    revo3_move_finger_wait(ctx.handle, ctx.slave_id, diagnostic_sequence[f].idx, diagnostic_sequence[f].target, 0.5f, 0.01f);
    // Stretch finger back
    revo3_move_finger_wait(ctx.handle, ctx.slave_id, diagnostic_sequence[f].idx, finger_zero, 0.5f, 0.01f);
  }

  std::printf("Testing: Thumb (smooth trajectory)...\n");
  float t_bend[5] = {0.0f, 30.0f, 30.0f, 30.0f, 30.0f};
  float t_zero[5] = {0.0f};
  revo3_move_thumb_wait(ctx.handle, ctx.slave_id, t_bend, 0.5f, 0.01f);
  revo3_move_thumb_wait(ctx.handle, ctx.slave_id, t_zero, 0.5f, 0.01f);

  std::printf("Restoring all joints back to Position control mode...\n");
  revo3_multi_joint_control(ctx.handle, ctx.slave_id, 0, zeros);
  revo3_sleep_ms(100);

  revo3_close(ctx);
  return 0;
}

File: python/revo3/hand_demo.py
#!/usr/bin/env python3
"""Revo3 hand demo: info, status, touch summary, and a small safe trajectory."""

import asyncio
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from common_imports import get_hw_type_name
from common_init import cleanup_context, parse_args_and_init


async def main():
    ctx, _, _ = await parse_args_and_init(sys.argv)
    if ctx is None:
        return 1

    try:
        client = ctx.ctx
        slave_id = ctx.slave_id

        info = await client.revo3_get_device_info(slave_id)
        print("=== Revo3 Python Hand Demo ===")
        print(f"Hardware: {get_hw_type_name(info.hardware_type)}")
        print(f"Serial: {info.serial_number}")
        print(f"Firmware: {info.firmware_version}")

        status = await client.revo3_get_motor_status_data(slave_id)
        print(f"Positions[0..4]: {[round(v, 2) for v in status.positions[:5]]}")
        print(f"Velocities[0..4]: {[round(v, 2) for v in status.velocities[:5]]}")
        print(f"Currents[0..4]: {[round(v, 2) for v in status.currents[:5]]}")

        try:
            enabled = await client.revo3_get_all_touch_modules_enabled(slave_id)
            if enabled:
                value_type = await client.revo3_get_touch_module_value_type(slave_id)
                summary = await client.revo3_get_touch_summary(slave_id)
                print(
                    f"Touch enabled=0x{enabled:03X}, "
                    f"value_type={int(value_type)}, summary[0..7]={summary[:8]}"
                )
            else:
                print("Touch modules are not enabled or not available.")
        except Exception as exc:
            print(f"Touch summary skipped: {exc}")

        zeros = [0.0] * 21
        grip_targets = [0.0] * 21
        for joint in (1, 2, 5, 6, 9, 10, 13, 14):
            grip_targets[joint] = 60.0
        for joint in (17, 18, 19, 20):
            grip_targets[joint] = 45.0

        # 1. Switch all joints to Impedance mode (4)
        print("Switching all joints to Impedance control mode (4) for smooth trajectory planning...")
        await client.revo3_multi_joint_control(slave_id, 4, zeros)
        await asyncio.sleep(0.1)

        # Step 1: Open all fingers to 0 deg with smooth trajectory
        print("Step 1: Open all fingers to 0 deg (Initial Stretch via revo3_move_hand_wait)...")
        await client.revo3_move_hand_wait(slave_id, zeros, 0.7, 0.025)

        # Step 2: Grip (Move whole hand to 60 deg with smooth trajectory)
        print("Step 2: Close whole hand to 60 deg (Grip via revo3_move_hand_wait)...")
        await client.revo3_move_hand_wait(slave_id, grip_targets, 0.7, 0.025)

        # Step 3: Open again (Return whole hand to 0 deg with smooth trajectory)
        print("Step 3: Open whole hand to 0 deg (Release via revo3_move_hand_wait)...")
        await client.revo3_move_hand_wait(slave_id, zeros, 0.7, 0.025)

        # Step 4: Test fingers one by one (Pinky -> Ring -> Middle -> Index -> Thumb)
        fingers = [
            ("Pinky (Little Finger)", 4, [0.0, 45.0, 45.0, 0.0]),
            ("Ring Finger", 3, [0.0, 45.0, 45.0, 0.0]),
            ("Middle Finger", 2, [0.0, 45.0, 45.0, 0.0]),
            ("Index Finger", 1, [0.0, 45.0, 45.0, 0.0]),
        ]

        for name, idx, finger_target in fingers:
            print(f"Testing: {name} (smooth trajectory)...")
            await client.revo3_move_finger_wait(slave_id, idx, finger_target, 0.5, 0.01)
            await client.revo3_move_finger_wait(slave_id, idx, [0.0] * 4, 0.5, 0.01)

        print("Testing: Thumb (smooth trajectory)...")
        await client.revo3_move_thumb_wait(slave_id, [0.0, 30.0, 30.0, 30.0, 30.0], 0.5, 0.01)
        await client.revo3_move_thumb_wait(slave_id, [0.0] * 5, 0.5, 0.01)

        # Restore all joints back to standard Position control mode (0)
        print("Restoring all joints back to Position control mode...")
        await client.revo3_multi_joint_control(slave_id, 0, zeros)
        await asyncio.sleep(0.1)

        return 0
    finally:
        await cleanup_context(ctx)


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))
